pkg/ipamd/datastore/data_store.go

```diff
@@ -781,20 +1036,36 @@ func (ds *DataStore) UnassignPodIPv4Address(ipamKey IPAMKey) (ip string, deviceN
 	if addr == nil {
 		ds.log.Warnf("UnassignPodIPv4Address: Failed to find sandbox %s",
 			ipamKey)
-		return "", 0, ErrUnknownPod
+		return nil, "", 0, false, ErrUnknownPod
 	}
 
 	ds.unassignPodIPv4AddressUnsafe(addr)
 	if err := ds.writeBackingStoreUnsafe(); err != nil {
 		// Unwind un-assignment
 		ds.assignPodIPv4AddressUnsafe(ipamKey, eni, addr)
-		return "", 0, err
+		return nil, "", 0, false, err
 	}
 	addr.UnassignedTime = time.Now()
+	if eni.IsPDenabled && addr.Prefix == "" {
+		ds.log.Infof("Prefix delegation is enabled and the IP is from secondary pool hence no need to update prefix pool")
+		ds.total--
+		mismatchedStore = true
+	} else if addr.Prefix != "" {
+		//Regular pod deletes for PD enabled and if pd is disabled but prefix is populated i.e, knob disable scenario
+		ds.log.Infof("Dump for prefix %v", addr.Prefix)
+		ds.log.Infof("DUMP - %v", eni.IPv4Prefixes[addr.Prefix])
+		deleteIPv4AddrfromPrefix(eni.IPv4Prefixes[addr.Prefix], addr)
+		//Remove the IP from eni DB
+		delete(eni.IPv4Addresses, addr.Address)
+		if !eni.IsPDenabled {
+			ds.log.Infof("Prefix delegation is disabled but IP belongs to prefix pool")
+			mismatchedStore = true
```

Comment:
I was initially confused what `mismatchedStore` meant.  Is `isSecondaryIP` equivalent and clearer?

Reply:
Yeah, make sense to have it as `isSecondaryIP`